File: LeetCode/Backtracking/python/permutations.py
from dataclasses import dataclass
from enum import Enum
from typing import List

class Result(Enum):
    PASS = "\033[92mPASS\033[00m"
    FAIL = "\033[91mFAIL\033[00m"

@dataclass
class permutationRecord:
    nums: List[int]
    expected: List[List[int]]
    
class Solution:
    def __init__(self):
        self.res = []
        self.visited = set()

    def permute(self, nums: List[int]) -> List[List[int]]:
        self.backtrack([], nums)
        return self.res
    
    def backtrack(self, path: List[int], nums: List[int]) -> None: 
        # Check if the length of our current values (path) is the same as the length of the nums array
        if len(path) == len(nums):
            # remember to perform a copy not a reference
            self.res.append(path[:])
            return
        
        # Assemble the permutation
        for i in range(len(nums)):
            # If we've seen this index before, just continue
            if i in self.visited:
                continue
            # Add the current index to our visited list
            self.visited.add(i)
            # Add the number to our path array
            path.append(nums[i])
            # backtrack
            self.backtrack(path, nums)
            # remove the current end of path
            path.pop()
            # remove current index from visited
            self.visited.remove(i)

    @staticmethod
    def testSolution(record: permutationRecord) -> None:
        print(f"input: nums: {record.nums}")
        print(f"expected: {record.expected}")
        res: List[List[int]] = Solution().permute(record.nums)
        print(f"result: {res}")
        print(Result.PASS.value if res == record.expected else Result.FAIL.value)

if __name__ == "__main__":
    records:  List[permutationRecord] = [
        permutationRecord([1,2,3], [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]),
        permutationRecord([0,1], [[0,1],[1,0]]),
        permutationRecord([1], [[1]])
    ]

    print("-" * 50)
    for i, record in enumerate(records):
        print(f"# Test case {i + 1}")
        Solution.testSolution(record)
        print("-" * 50)

File: LeetCode/Backtracking/java/permutations/permutations.java
package LeetCode.Backtracking.java.permutations;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

enum testResult {
    PASS("\u001B[32mPASS\u001B[0m"),
    FAIL("\u001B[31mFAIL\u001B[0m");

    private final String value;

    testResult(String value) {
        this.value = value;
    }

    public String getValue() {
        return this.value;
    }

}

record permutationsRecord(int[] nums, List<List<Integer>> expected) {
}

public class permutations {
    List<List<Integer>> res = new ArrayList<>();
    Set<Integer> visited = new HashSet<>();

    public List<List<Integer>> permute(int[] nums) {
        this.backtrack(new ArrayList<>(), nums);

        return this.res;
    }

    private void backtrack(List<Integer> path, int[] nums) {
        // Check if the length of our current values (path) is the same as the length of
        // the nums array
        if (path.size() == nums.length) {
            // remember to perform a copy not a reference
            this.res.add(new ArrayList<>(path));
            return;
        }

        // Assemble the permutation
        for (int i = 0; i < nums.length; i++) {
            // If we've seen this index before, just continue
            if (this.visited.contains(i)) {
                continue;
            }

            this.visited.add(i);

            path.add(nums[i]);

            this.backtrack(path, nums);

            path.removeLast();

            this.visited.remove(i);
        }
    }

    public static void main(String[] args) {
        permutationsRecord[] records = new permutationsRecord[] {
                new permutationsRecord(new int[] { 1, 2, 3 }, List.of(List.of(1, 2, 3), List.of(1, 3, 2),
                        List.of(2, 1, 3), List.of(2, 3, 1), List.of(3, 1, 2), List.of(3, 2, 1))),
                new permutationsRecord(new int[] { 0, 1 }, List.of(List.of(0, 1), List.of(1, 0))),
                new permutationsRecord(new int[] { 1 }, List.of(List.of(1)))
        };

        int i = 1;
        for (permutationsRecord record : records) {
            System.out.println("# Test case " + i++);
            permutations.testSolution(record);
            System.out.println("-".repeat(50));
        }
    }

    private static void testSolution(permutationsRecord record) {
        System.out.println("input: nums: " + Arrays.toString(record.nums()));
        System.out.println("expected: " + record.expected());
        List<List<Integer>> res = new permutations().permute(record.nums());
        System.out.println("result: " + res);
        System.out.println(res.equals(record.expected()) ? testResult.PASS.getValue() : testResult.FAIL.getValue());
    }
}
